source/extensions/filters/network/zookeeper_proxy/zookeeper_filter.cc

```diff
@@ -0,0 +1,214 @@
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_filter.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/common/logger.h"
+
+#include "extensions/filters/network/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ZooKeeperProxy {
+
+ZooKeeperFilterConfig::ZooKeeperFilterConfig(const std::string& stat_prefix, Stats::Scope& scope)
+    : scope_(scope), stat_prefix_(stat_prefix), stats_(generateStats(stat_prefix, scope)) {}
+
+ZooKeeperFilter::ZooKeeperFilter(ZooKeeperFilterConfigSharedPtr config)
+    : config_(std::move(config)) {}
+
+void ZooKeeperFilter::initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) {
+  read_callbacks_ = &callbacks;
+}
+
+Network::FilterStatus ZooKeeperFilter::onData(Buffer::Instance& data, bool) {
+  doDecode(data);
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus ZooKeeperFilter::onWrite(Buffer::Instance&, bool) {
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus ZooKeeperFilter::onNewConnection() { return Network::FilterStatus::Continue; }
+
+void ZooKeeperFilter::doDecode(Buffer::Instance& buffer) {
+  // Clear dynamic metadata.
+  envoy::api::v2::core::Metadata& dynamic_metadata =
+      read_callbacks_->connection().streamInfo().dynamicMetadata();
+  auto& metadata =
+      (*dynamic_metadata.mutable_filter_metadata())[NetworkFilterNames::get().ZooKeeperProxy];
+  metadata.mutable_fields()->clear();
+
+  if (!decoder_) {
+    decoder_ = createDecoder(*this);
+  }
+
+  try {
+    decoder_->onData(buffer);
+  } catch (EnvoyException& e) {
+    ENVOY_LOG(info, "zookeeper_proxy: decoding error: {}", e.what());
+    config_->stats_.decoder_error_.inc();
+  }
+}
+
+DecoderPtr ZooKeeperFilter::createDecoder(DecoderCallbacks& callbacks) {
+  return std::make_unique<DecoderImpl>(callbacks);
+}
+
+void ZooKeeperFilter::setDynamicMetadata(const std::string& key, const std::string& value) {
+  setDynamicMetadata({{key, value}});
+}
+
+void ZooKeeperFilter::setDynamicMetadata(
+    const std::vector<std::pair<const std::string, const std::string>>& data) {
+  envoy::api::v2::core::Metadata& dynamic_metadata =
+      read_callbacks_->connection().streamInfo().dynamicMetadata();
+  ProtobufWkt::Struct metadata(
+      (*dynamic_metadata.mutable_filter_metadata())[NetworkFilterNames::get().ZooKeeperProxy]);
+  auto& fields = *metadata.mutable_fields();
+
+  for (const auto& pair : data) {
+    auto val = ProtobufWkt::Value();
+    val.set_string_value(pair.second);
+    fields.insert({pair.first, val});
+  }
+
+  read_callbacks_->connection().streamInfo().setDynamicMetadata(
```

Comment:
For my education: how does this makes it way into the access logs?

Reply:
This is not implement yet; I am planning on doing it like the mongo proxy filter does it... 

Though, what do you think of leaving this for a follow-up PR?

Reply:
Resolving, assuming it'll be ok to do this in the next PR (it's marked as not implemented in the proto config fil...).